pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,475 @@
+					}
+
+					num_joining = changes.collators.inc.len().saturated_into();
+					num_leaving = changes.collators.out.len().saturated_into();
+
+					// Apply collator set changes AFTER rewarding
+					for leaving in changes.collators.out.iter() {
+						Self::do_exit_collator(leaving)?;
+					}
+					for joining in changes.collators.inc.iter() {
+						Self::do_init_collator(joining)?;
+					}
+
+					// Apply session changes
+					session_data.collator_reward = changes
+						.collator_reward
+						.unwrap_or(session_data.collator_reward);
+					session_data.total_reward =
+						changes.total_reward.unwrap_or(session_data.total_reward);
+					session_data.collator_count = changes
+						.collator_count
+						.unwrap_or(session_data.collator_count);
+
+					Self::deposit_event(Event::NewSession {
+						total_reward: session_data.total_reward,
+						collator_reward: session_data.collator_reward,
+						last_changes: mem::take(changes),
+					});
+
+					Ok(())
+				})
+			})
+		})
+		.map_err(|e| {
+			log::error!("Failed to advance block rewards session: {:?}", e);
+		})
+		.ok();
+	}
+}
+
+impl<T: Config> sp_runtime::BoundToRuntimeAppPublic for Pallet<T> {
+	type Public = T::AuthorityId;
+}
+
+// Should be instantiated after the original SessionHandler such that current and queued collators are up-to-date for the current session.
+impl<T: Config> OneSessionHandler<T::AccountId> for Pallet<T> {
+	type Key = T::AuthorityId;
+
+	fn on_genesis_session<'a, I: 'a>(_validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+	{
+		// Handled in genesis builder.
+	}
+
+	fn on_new_session<'a, I: 'a>(_: bool, validators: I, queued_validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+	{
+		// MUST be called before updating collator set changes.
+		// Else the timing is off.
+		Self::do_advance_session();
+		let current = validators
+			.map(|(acc_id, _)| acc_id.clone())
+			.collect::<Vec<_>>();
+		let next = queued_validators
+			.map(|(acc_id, _)| acc_id.clone())
+			.collect::<Vec<_>>();
+
+		// Prepare collator set changes for next session.
+		if current != next {
+			// Prepare for next session
+			NextSessionChanges::<T>::mutate(
+				|SessionChanges {
+				     collators,
+				     collator_count,
+				     ..
+				 }| {
+					let inc = next
+						.clone()
+						.into_iter()
+						.filter(|n| !current.iter().any(|curr| curr == n))
+						.collect::<Vec<_>>();
+					let out = current
+						.clone()
+						.into_iter()
+						.filter(|curr| !next.iter().any(|n| n == curr))
+						.collect::<Vec<_>>();
+					collators.inc = BoundedVec::<_, T::MaxCollators>::truncate_from(inc);
+					collators.out = BoundedVec::<_, T::MaxCollators>::truncate_from(out);
+
+					*collator_count = Some(next.len().saturated_into::<u32>());
+				},
+			);
+		}
+	}
+
+	fn on_before_session_ending() {
+		// we don't care.
+	}
+
+	fn on_disabled(_validator_index: u32) {
+		// we don't care.
```

Comment:
What means disabling from a session perspective?

Reply:
In the context of AuRa/Babe, it prevents such collators to author a block , see https://github.com/paritytech/substrate/pull/9414. However, slots assigned to disabled authorities are just skipped in AuRa e.g. no block is built. Thus, there is no incentive to disable a collator IMO.

In the context of a session, we could unstake these collators. However, their session keys are still part of `QueuedKeys`. If we wanted to handle this scenario, more checks would have to be introduced to `on_new_session`